### ISSUE TITLE
Segmentation fault when running the "ninja check-all" second time.

### ISSUE THREAD

**Issue by MaggieYingYi** [SNSystems/llvm-project-prepo#41]
Following the instructions mentioned in [the Wiki page](https://github.com/SNSystems/llvm-project-prepo/wiki/Build-LLVM-with-the-Repo-Compiler), the LLVM project is built with the release configuration using the master branch (https://github.com/SNSystems/llvm-project-prepo/commit/4602820f34f7d3c8340c46c526b0312c96d8b3ed).

The build and all tests passed on the first time. The second time built fine. A segmentation fault is met when running tests with the following command:
`$ cd /home/maggie/github/llvm-project-prepo/clang/bindings/python && /usr/bin/cmake -E env CLAN...H=/home/maggie/github/llvm-project-prepo/build_repo_release/lib /usr/bin/python2.7 -m unittest discover`

**Comment by MaggieYingYi** [comment]
After debugging the core dump file, I found that the issue is caused by calling the function of `_GLOBAL__sub_I_EnumTables.cpp`, which is  defined in the file [EnumTables.cpp](llvm-project-prepo\llvm\lib\DebugInfo\PDB\Native\EnumTables.cpp).

If we compiled the `EnumTables.cpp` file twice and compared the compilation members between the two ticket files. I found the compilation members of `llvm.global_ctors`  and `_GLOBAL__sub_I_EnumTables.cpp` only exist in the second time build. 

The `llvm.global_ctors` array contains a list of constructor functions, priorities, and an associated global or function, such as the `_GLOBAL__sub_I_EnumTables.cpp`. The `_GLOBAL__sub_I_EnumTables.cpp` will be called when the module is loaded. When running the check-all command for the second time, the function `_GLOBAL__sub_I_EnumTables.cpp` crashed.

To fix the issue, `llvm.global_ctors`  and `_GLOBAL__sub_I_EnumTables.cpp` should not exist in the ticket file for the second time build.

**Comment by MaggieYingYi** [comment]
The issue is related with the `RepoPruning` pass and the `elim-avail-extern` pass (eliminate available externally globals).

The `RepoPruning` pass removes unchanged global objects (GOs) whose definitions already have binary representations in the repository by borrowing the existing `available_externally` linkage type. GOs with “available_externally” linkage are escaped all optimization passes. They exist to allow inlining and other optimizations to take place given knowledge of the definition of the global, which is known to be somewhere outside the module. If a GO is pruned, its linkage type is changed to the `available_externally` linkage type to avoid the later phases of the compiling process. 

**In the `EnumTables.cpp` file, let’s consider the `llvm.global_ctors` global variable.**

The `llvm.global_ctors `global variable calls the function `_GLOBAL__sub_I_EnumTables.cpp`. The function `_GLOBAL__sub_I_EnumTables.cpp` calls the function `__cxx_global_var_init`. Then, the function `__cxx_global_var_init` calls `_ZN4llvm9EnumEntryItEC1ENS_9StringRefEt` and `_ZN4llvm9StringRefC1EPKc`.

**Compile the EnumTables.cpp file (first-time), we find:**

1. After `RepoMetadataGenerationPass`, the digests for `llvm.global_ctors`, `_GLOBAL__sub_I_EnumTables.cpp`, `__cxx_global_var_init, _ZN4llvm9StringRefC1EPKc`, `_ZN4llvm9EnumEntryItEC1ENS_9StringRefEt` and `_ZL22OMFSegMapDescFlagNames` are generated.
2. After `RepoPruningPass`, there is **NO** pruning happen since there is **NO** database.
3. After `Function Integration/Inlining` pass, the functions `_ZN4llvm9StringRefC1EPKc` and `_ZN4llvm9EnumEntryItEC1ENS_9StringRefEt` are inlined into the function `__cxx_global_var_init`. Then, the function `__cxx_global_var_init` is inlined into the function `_GLOBAL__sub_I_EnumTables.cpp`. The `_ZN4llvm9StringRefC1EPKc`,  `_ZN4llvm9EnumEntryItEC1ENS_9StringRefEt` and `__cxx_global_var_init` functions have been removed. 
After ininling, the function `_GLOBAL__sub_I_EnumTables.cpp` has an external fix-up to the global variable `_ZL22OMFSegMapDescFlagNames`.
4. After `Eliminate Available Externally Globals` pass, there is **NO** change since there is **NO** `available_externally` linkage type for all global objects.
5. After `Global Variable Optimizer` pass, the `llvm.global_ctors` and `_GLOBAL__sub_I_EnumTables.cpp` have been removed.
6. During repo object writer, `_ZL22OMFSegMapDescFlagNames` is written into the database.

**To recompile the file with existing database (all code fragments are retained within the database).**

1. After `RepoMetadataGenerationPass`, the IR code is the same as the first-time compilation - all digests are generated.
2. After `RepoPruningPass`, `_ZL22OMFSegMapDescFlagNames` is pruned since it was stored in the database. Its linkage type is changed to the `available_externally` linkage type.
3. After `Function Integration/Inlining` pass, the `_ZN4llvm9EnumEntryItEC1ENS_9StringRefEt`, `_ZN4llvm9StringRefC1EPKc` and `__cxx_global_var_init` functions have been inlined and removed, which is the same as the first-time compilation.
4. After `Eliminate Available Externally Globals` pass, the `_ZL22OMFSegMapDescFlagNames` has been changed from `definition` to `declaration` since its linkage type has been changed to the `available_externally` linkage type in the step 2.
5. After `Global Variable Optimizer` pass, the `llvm.global_ctors` and `_GLOBAL__sub_I_EnumTables.cpp` have **NOT** been removed since there is **NO** definition for the global variable '_ZL22OMFSegMapDescFlagNames'.
6. During Repo object writer, the fragments of `_GLOBAL__sub_I_EnumTables.cpp` and `llvm.global_ctors` are written into the database. This is caused the issue #41.

**Comment by MaggieYingYi** [comment]
The `Eliminate Available Externally Globals` pass is designed to eliminate available external global definitions from the program. If the global objects have available_externally linkage types, they are turned from definitions into declarations. The initial GO’s linkage type is saved in the TicketNode metadata for the Repo target. There are two possible solutions:

1. Teach the `Eliminate Available Externally Globals` pass about the repo.
2. Add a new `Repo Eliminate Available Externally Globals` pass (named `RepoEliminateAvailableExternallyPass`) and a new clang flag (`-frepo`).

    Currently the `Eliminate Available Externally Globals` pass is enabled by the following code:
    ```
    if (OptLevel > 1 && !PrepareForLTO && !PrepareForThinLTO)
        MPM.add(createEliminateAvailableExternallyPass());
    ```
   Changed to:

    ```
    if (OptLevel > 1 && !PrepareForLTO && !PrepareForThinLTO && !Repo)
        MPM.add(createEliminateAvailableExternallyPass());
    if (OptLevel > 1 && Repo)
        MPM.add(createRepoEliminateAvailableExternallyPass());
    ```

The `RepoEliminateAvailableExternallyPass` is only used when targeting on the repo and it has the same functionality as the `EliminateAvailableExternallyPass`.

We've worked hard to avoid teaching specific optimizations about the repo (method 1) because the optimization passes will change in the future. However, the second method changes the pass manage builder, clang (the new flag `-frepo`) and adds the new repo pass, which is similar to the `Eliminate Available Externally Globals` pass (redundant codes). Therefore, I decide to use the first method to fix the issue.
